pkg/security/security_profile/profile/profile_dir.go

```diff
@@ -0,0 +1,384 @@
+	"github.com/DataDog/datadog-agent/pkg/security/seclog"
+	"github.com/DataDog/datadog-agent/pkg/security/utils"
+)
+
+var (
+	debounceDelay        = 5 * time.Second
+	newFileDebounceDelay = 2 * time.Second
+)
+
+const profileExtension = ".profile"
+
+// make sure the DirectoryProvider implements Provider
+var _ Provider = (*DirectoryProvider)(nil)
+
+type profileFSEntry struct {
+	path    string
+	version string
+}
+
+// DirectoryProvider is a ProfileProvider that fetches Security Profiles from the filesystem
+type DirectoryProvider struct {
+	sync.Mutex
+	directory      string
+	watcherEnabled bool
+
+	// attributes used by the inotify watcher
+	cancelFnc         func()
+	watcher           *fsnotify.Watcher
+	watchedFiles      []string
+	newFilesDebouncer *debouncer.Debouncer
+	newFiles          map[string]int
+	newFilesLock      sync.Mutex
+
+	// we use a debouncer to forward new profiles to the profile manager in order to prevent a deadlock
+	workloadSelectorDebouncer *debouncer.Debouncer
+	onNewProfileCallback      func(selector cgroupModel.WorkloadSelector, profile *proto.SecurityProfile)
+
+	// selectors is used to select the profiles we currently care about
+	selectors []cgroupModel.WorkloadSelector
+	// profileMapping is an in-memory mapping of the profiles currently on the file system
+	profileMapping map[cgroupModel.WorkloadSelector]profileFSEntry
+}
+
+// NewDirectoryProvider returns a new instance of DirectoryProvider
+func NewDirectoryProvider(directory string, watch bool) (*DirectoryProvider, error) {
+	// check if the provided directory exists
+	if fi, err := os.Stat(directory); err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(directory, 0750); err != nil {
+				return nil, fmt.Errorf("can't create security profiles directory `%s`: %w", directory, err)
+			}
+		} else {
+			return nil, fmt.Errorf("can't load security profiles from `%s`: %w", directory, err)
+		}
+	} else {
+		// check the permissions of the folder
+		if fi.Mode().Perm() != 0750 {
+			return nil, fmt.Errorf("the security profiles local directory doesn't have the correct access mode: expected 0750, got %o", fi.Mode().Perm())
+		}
+	}
+
+	dp := &DirectoryProvider{
+		directory:      directory,
+		watcherEnabled: watch,
+		profileMapping: make(map[cgroupModel.WorkloadSelector]profileFSEntry),
+		newFiles:       make(map[string]int),
+	}
+	dp.workloadSelectorDebouncer = debouncer.New(debounceDelay, dp.onNewProfileDebouncerCallback)
+	dp.newFilesDebouncer = debouncer.New(newFileDebounceDelay, dp.onHandleFilesFromWatcher)
+
+	return dp, nil
+}
+
+// Start runs the directory provider
+func (dp *DirectoryProvider) Start(ctx context.Context) error {
+	dp.workloadSelectorDebouncer.Start()
+	dp.newFilesDebouncer.Start()
+
+	// start by loading the profiles in the configured directory
+	if err := dp.loadProfiles(); err != nil {
+		return fmt.Errorf("couldn't scan the security profiles directory: %w", err)
```

Comment:
Is there a race if we receive a profile after loadProfiles() but before adding the fsnotify watcher?
Unlikely, but we may want to swap these two to avoid it